### PR TITLE
Fix: QuestGraph

### PR DESCRIPTION
**Pull request by alexandrecorlet**
Introduction:
Hello, In this PR, a series of essential corrections and improvements were implemented for the QuestGraph class, providing a more robust and efficient functionality. The main fix focused on the isReachable functionality, which previously did not consider intermediate nodes to reach the destination node. With the Depth-First Search (DFS) strategy now implemented, the method properly considers these nodes, significantly enhancing the accuracy of accessibility checks.

Details of the Fixes:
Before the fix, the isReachable method did not consider the possibility of intermediate nodes to reach the destination node. This could lead to false negatives, indicating that certain nodes were not reachable when they actually were. With the implementation of the Depth-First Search (DFS) strategy, the functionality now properly considers all possible paths to determine the accessibility of the destination node.

Additionally, to ensure the robustness of the fix, additional test cases were created, covering a variety of scenarios to validate the accuracy and reliability of the isReachable method after the adjustment.

Refactorings in the Class:
In addition to the specific fixes, this PR also included a series of refactorings in the QuestGraph class, aiming to improve its readability, maintainability, and overall performance. The most significant of these refactorings occurred in the init() and update() methods, where switches were replaced by a polymorphism-based strategy.

This polymorphic approach allows for more efficient computation of the quest task type and its corresponding treatment, resulting in a considerable reduction in "bad smells" and simplifying the code execution flow.

Conclusion:
In summary, this PR not only addresses a crucial issue in the isReachable functionality of the QuestGraph class but also introduces significant improvements in its internal structure and organization. With these changes, we can ensure a more consistent and reliable experience for users who rely on this class to manage quests in our system.

### CODE REVIEW

**Comment by hdescottes** [comment]
Hello, thank you for your contribution and your interest in this project ! :)
I did not look at that quest system in a while and it needed some serious dusting !
I will have a look as quickly as possible. In the meantime, please check the pipeline and try to fix the issues :)

**Comment by alexandrecorlet** [comment]
> Hello, thank you for your contribution and your interest in this project ! :) I did not look at that quest system in a while and it needed some serious dusting ! I will have a look as quickly as possible. In the meantime, please check the pipeline and try to fix the issues :)

Sure! I'll review the pipeline and try to fix the issues! :)

File: core/src/main/java/com/gdx/game/quest/QuestGraph.java [hunk 1]
```diff
@@ -28,6 +25,11 @@ public class QuestGraph {
     private int goldReward;
     private int xpReward;
 
+    public QuestGraph() {
```
Comment (hdescottes) [inline core/src/main/java/com/gdx/game/quest/QuestGraph.java:28]: QuestGraph should not be initialized with a constructor (except for test)
If you want to initialize some variables, do it inside the class itself

**Review by hdescottes — CHANGES_REQUESTED** [review]
There is an issue in the quest completion system with what you did.
I can pick up a quest, retrieve teh items needed, but i cannot complete it.

For example :

This PNJ gives you a simple quest, you have to find her baby (in the village)
image
Once you have it, go back to that same PNJ and say "yes", you should have enough XP to level up and this window should appear
image

This does not happen with the current code in your branch :/

**Comment by alexandrecorlet** [comment]
> There is an issue in the quest completion system with what you did. I can pick up a quest, retrieve teh items needed, but i cannot complete it.
> 
> For example :
> 
> This PNJ gives you a simple quest, you have to find her baby (in the village) image Once you have it, go back to that same PNJ and say "yes", you should have enough XP to level up and this window should appear image
> 
> This does not happen with the current code in your branch :/ (it seems like the test failling on the pipeline is related to that :), once you fix it it should work)

Hi,

Apparently, the problem was because I've accidentally changed the behavior of the methods `updateQuestForReturn` and `isQuestTaskAvailable` when I refactored them. I've fixed the issue by reverting both methods to their previous implementation. It seems that game is working properly now. :)

**Comment by hdescottes** [comment]
> > There is an issue in the quest completion system with what you did. I can pick up a quest, retrieve teh items needed, but i cannot complete it.
> > For example :
> > This PNJ gives you a simple quest, you have to find her baby (in the village) image Once you have it, go back to that same PNJ and say "yes", you should have enough XP to level up and this window should appear image
> > This does not happen with the current code in your branch :/ (it seems like the test failling on the pipeline is related to that :), once you fix it it should work)
> 
> Hi,
> 
> Apparently, the problem was because I've accidentally changed the behavior of the methods `updateQuestForReturn` and `isQuestTaskAvailable` when I refactored them. I've fixed the issue by reverting both methods to their previous implementation. It seems that game is working properly now. :)

nice i will review that asap :)

**Review by hdescottes — CHANGES_REQUESTED** [review]
The quest completion system is fine now :)

But i have spotted something weird. NPCs should always have some random conversation like "Hi", "Greetings stranger", etc... if they don't have quests for the player. This does not work anymore :/ (https://github.com/hdescottes/GdxGame/commit/2db9e19004b1cd2d0f048976f2bb43b93b78f247)
Make sure to write a test about this behaviour in order to not have this error again.

**Comment by alexandrecorlet** [comment]
> The quest completion system is fine now :)
> 
> But i have spotted something weird. NPCs should always have some random conversation like "Hi", "Greetings stranger", etc... if they don't have quests for the player. This does not work anymore :/ ([2db9e19](https://github.com/hdescottes/GdxGame/commit/2db9e19004b1cd2d0f048976f2bb43b93b78f247)) Fix this and make sure to write a test about this behaviour in order to not have this error again.

Hi, Hugo. How are you? :)

Did you spot this strange behavior on the `master` or  `fix/questGraph` branch? I've tried to reproduce the problem you mentioned on the `fix/questGraph`, and the NPC behaved normally:

Screen Shot 2024-05-19 at 14 59 32

I've also checked the game logs, and they're ok too:

Screen Shot 2024-05-19 at 14 59 39

So, after investigating a little bit, I've switched to the `master` branch, and tried to reproduce the error you mentioned there. This is the game log when I tried to speak with a NPC on the `master` branch:

Screen Shot 2024-05-19 at 14 58 05

Apparently, the error is occurring because the conversation file was deleted.

File: core/src/main/java/com/gdx/game/quest/QuestGraph.java [hunk 2]
```diff
@@ -225,119 +189,46 @@ public void update(MapManager mapMgr) {
...
-                    break;
-                case RETURN:
-                    break;
-                case DISCOVER:
-                    break;
-            }
+    private QuestTaskHandler getHandlerForQuestType(QuestTask.QuestType questType) {
+        switch (questType) {
+            case FETCH:
+                return new FetchQuestTaskHandler();
```
[the remark this answers is not in the document]
Comment (alexandrecorlet) [reply core/src/main/java/com/gdx/game/quest/QuestGraph.java:212]: Ty!! :D

**Comment by hdescottes** [comment]
Huh, that's weird, because i can see the file in both masters branch (mine and yours).
But i can't see it when i copy your repo with your branch
WEIRD

I will merge the PR that it seems that is something that somehow i did while copy your repo?

**Comment by alexandrecorlet** [comment]
Yeah, `master` branch is missing the `conversation_courtesy.json` file. IDK how this happened lol. REALLY WEIRD ;-;

Anyway, I'm glad I could contribute to your project! Thanks for the opportunity and hopefully we can work together again sometime! Thx again and happy coding, friend! :)
